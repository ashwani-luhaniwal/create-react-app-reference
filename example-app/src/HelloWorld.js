import React, { Component } from 'react';
import './HelloWorld.css';

/**
 * props are properties inside of React component that are passed in from somewhere
 * generally from parent.
 * @param {props} param0 
 */
/*const HelloWorld = (props) => {
    return (<div className="HelloWorld">Hello {props.name}</div>);
};
export default HelloWorld;*/

// HelloWorld component extends functionality of Component Class 
class HelloWorld extends Component {
    constructor(props) {
        // calls parent's constructor using super(props).
        super(props);
        // set property on class called "state"
        this.state = { greeting: 'Hello' };
        // Tells JavaScript, any time you see a reference to "this" inside of the frenchify
        // function, please refer to ME.
        this.frenchify = this.frenchify.bind(this);
    }
    frenchify() {
        // setState only changes keys that are specified inside of state, it doesn't replace
        // the entire state.
        this.setState({ greeting: 'Bonjour' });
    }
    render() {
        return (
            <div className="HelloWorld">
                {this.state.greeting} {this.props.name}!
                <br />
                <button onClick={this.frenchify}>Frenchify!</button>
            </div>
        );
    }
}
export default HelloWorld;