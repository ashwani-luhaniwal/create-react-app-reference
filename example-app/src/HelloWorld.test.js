import React from 'react';
import { shallow } from 'enzyme';
// "renderer" object from "react-test-renderer" to build our snapshot test that Jest will use.
// This functionality grabs a rendered "snapshot" of your component tree and compares it
// against a stored snapshot.
import renderer from 'react-test-renderer';

import HelloWorld from './HelloWorld';

/**
 * A describe function tells Jest that our tests describe the behavior and functionality of
 * particular component, concept or suite of tests. You can either describe a component using
 * its name or you can just provide a string to a describe function, like as follows: 
 * 
 *  describe('My Test', () => { ... });
 */
describe(HelloWorld, () => {
    // Add the rest of our tests here
    const name = 'Person';
    /**
     * jest.fn() is special syntax to create a mock function. A mock is a facsimile function:
     * it doesn't actually do the work that it stands in place of; it merely simulates it,
     * records attempts to call it, and optionally looks for arguments or fakes return 
     * values. 
     */
    const mockRemoveGreeting = jest.fn();
    // shallow render
    /**
     * A shallow render is a simulated render of a component tree. It allows for deeper
     * inspection then the snapshot level, such as looking for specific DOM elements and 
     * their contents, as well as simulating user interaction.
     */
    const component = shallow(
        <HelloWorld name={name} removeGreeting={mockRemoveGreeting} />
    );

    /**
     * We start off by calling "renderer.create", and passing to that function the JSX that
     * we want to build into our new snapshot. Then we take that snapshot and turn it into
     * a JSON representation of the tree. Finally using Jest, we tell it that we are 
     * expecting JSON representation of our snapshot to match the stored snapshot that we 
     * already have. This will tell us if someone starts making fundamental changes to our
     * component without updating the tests!
     */
    it('renders and matches our snapshot', () => {
        const component = renderer.create(
            <HelloWorld name="Person" />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('contains the supplied name', () => {
        // It grabs the text inside of rendered component, and setup an expectation that
        // the text should contain the name that we passed to the component via props.
        expect(component.text()).toContain(name);
    });

    it('modifies the greeting when frenchify button is clicked', () => {
        component.find('button.frenchify').simulate('click');
        expect(component.text()).toContain('Bonjour');
    });

    it('calls the passed in removeGreeting function when remove button is clicked', () => {
        component.find('button.remove').simulate('click');
        expect(mockRemoveGreeting).toBeCalled();
    });
});