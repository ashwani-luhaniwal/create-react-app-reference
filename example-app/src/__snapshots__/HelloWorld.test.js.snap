// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class HelloWorld extends _react.Component {
    constructor(props) {
        // calls parent's constructor using super(props).
        super(props);
        // set property on class called "state"
        this.state = { greeting: 'Hello' };
        // Tells JavaScript, any time you see a reference to "this" inside of the frenchify
        // function, please refer to ME.
        this.frenchify = this.frenchify.bind(this);
        this.removeGreeting = this.removeGreeting.bind(this);
    }
    frenchify() {
        // setState only changes keys that are specified inside of state, it doesn't replace
        // the entire state.
        this.setState({ greeting: 'Bonjour' });
    }
    removeGreeting() {
        this.props.removeGreeting(this.props.name);
    }
    render() {
        return (
            _react2.default.createElement('div', { className: 'HelloWorld', __source: { fileName: _jsxFileName, lineNumber: 36 }, __self: this },
                this.state.greeting, ' ', this.props.name, '!',
                _react2.default.createElement('br', { __source: { fileName: _jsxFileName, lineNumber: 38 }, __self: this }),
                _react2.default.createElement('button', { className: 'frenchify', onClick: this.frenchify, __source: { fileName: _jsxFileName, lineNumber: 39 }, __self: this }, 'Frenchify!'),
                _react2.default.createElement('br', { __source: { fileName: _jsxFileName, lineNumber: 40 }, __self: this }),
                _react2.default.createElement('button', { className: 'remove', onClick: this.removeGreeting, __source: { fileName: _jsxFileName, lineNumber: 41 }, __self: this }, 'Remove Me!')));


    }} renders and matches our snapshot 1`] = `
<div
  className="HelloWorld"
>
  Hello
   
  Person
  !
  <br />
  <button
    className="frenchify"
    onClick={[Function]}
  >
    Frenchify!
  </button>
  <br />
  <button
    className="remove"
    onClick={[Function]}
  >
    Remove Me!
  </button>
</div>
`;
